test: don't use deprecated way of capturing logs

daipom opened the pull request:
# Background

The next [td-agent](https://github.com/fluent/fluent-package-builder) will embed Ruby 3.2.

I was checking each embedded plugin and noticed that this plugin has no GitHub CI.

I added CI to my forked repository, but I found the tests are unstable on Windows and macOS.

* https://github.com/daipom/fluent-plugin-flowcounter-simple/actions/runs/4239789791

So, I created this PR first.
Following this, I would like to create a PR to add GitHub CI.

# About this fix

We should use `Driver::logs` instead of `capture_log` to fix unstable tests.

`capture_log` is defined by this plugin here.

https://github.com/fluent-plugins-nursery/fluent-plugin-flowcounter-simple/blob/5081e88f97156e019115cab2b4cd810c28ca2ed0/test/helper.rb#L12-L28

However, this is for older versions as well as the following, and this makes tests unstable.

* https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/lib/fluent/test/helpers.rb#L110-L122

# Reproduce failing tests

Windows seems to have the highest probability of failure.

* my forked repository's CI: https://github.com/daipom/fluent-plugin-flowcounter-simple/actions/runs/4239789791
* On my Windows10 Home machine
  * Ruby 3.2
  * Each test sometimes succeeds and sometimes does not.
  ```console
  $ bundle
  $ bundle exec rake
  C:/Ruby32-x64/bin/ruby.exe -w -I"lib;lib;test" C:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb "test/plugin/test_filter_flowcounter_simple.rb" "test/plugin/test_out_flowcounter_simple.rb"
  C:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/bundler-2.3.0/lib/bundler/vendor/thor/lib/thor/error.rb:105: warning: constant DidYouMean::SPELL_CHECKERS is deprecated
  Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead.
  C:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/serverengine-2.3.1/lib/serverengine/socket_manager_win.rb:140: warning: assigned but unused variable - type
  C:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/cool.io-1.7.1/lib/iobuffer_ext.so: warning: method redefined; discarding old initialize
  C:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/cool.io-1.7.1/lib/iobuffer_ext.so: warning: method redefined; discarding old clear
  C:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/cool.io-1.7.1/lib/iobuffer_ext.so: warning: method redefined; discarding old empty?
  C:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/cool.io-1.7.1/lib/iobuffer_ext.so: warning: method redefined; discarding old write
  C:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/fluentd-1.15.3-x64-mingw-ucrt/lib/fluent/plugin_helper.rb:46: warning: method redefined; discarding old inherited
  C:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/fluentd-1.15.3-x64-mingw-ucrt/lib/fluent/plugin_helper.rb:46: warning: previous definition of inherited was here
  Loaded suite C:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/rake-13.0.6/lib/rake/rake_test_loader
  Started
  F
  ==============================================================================================================================================================================================================================================
        assert { out.include?("count:20") }
                 |   |
                 |   false
                 ""
  C:/Users/reang/Documents/work/fluentd/fluentdPlugins/fluent-plugin-flowcounter-simple/test/plugin/test_filter_flowcounter_simple.rb:30:in `test_filter'
       27:     d = create_driver
       28:     filtered, out = filter(d, msgs)
       29:     assert_equal msgs, filtered
    => 30:     assert { out.include?("count:20") }
       31:   end
       32:
       33:   private
  ==============================================================================================================================================================================================================================================....
  Finished in 0.5054875 seconds.
  ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------5 tests, 9 assertions, 1 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
  ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------9.89 tests/s, 17.80 assertions/s
  Command failed with status (1): [ruby -w -I"lib;lib;test" C:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb "test/plugin/test_filter_flowcounter_simple.rb" "test/plugin/test_out_flowcounter_simple.rb" ]

  Tasks: TOP => default => test
  (See full trace by running task with --trace)
  ```

kou commented on test/plugin/test_filter_flowcounter_simple.rb:
```diff
@@ -25,9 +25,9 @@ def test_filter
+    filtered, logs = filter(d, msgs)
     assert_equal msgs, filtered
-    assert { out.include?("count:20") }
+    assert { logs.any? { |log| log.include?("count:20") } }
```
Could you show `pp logs`?
I want to think about better assertion with the output.

daipom replied:
Sure!
For all tests, the number of logs was 1, and it is tab-delimited.

```console
$ bundle exec rake test TESTOPTS='-v'
...
FlowCounterSimpleFilterTest:
  test_filter:
    ["2023-04-05 11:01:34 +0900 [info]: plugin:filter_flowcounter_simple\tcount:20\tindicator:num\tunit:second\n"]
    .: (0.156808)
FlowCounterSimpleOutputTest:
  test_byte:
    ["2023-04-05 11:01:35 +0900 [info]: plugin:out_flowcounter_simple\tcount:3330\tindicator:byte\tunit:second\n"]
    .: (0.133296)
  test_comment:
    ["2023-04-05 11:01:35 +0900 [info]: plugin:out_flowcounter_simple\tcount:3\tindicator:num\tunit:second\tcomment:foobar\n"]
    .: (0.113249)
  test_configure:
    .: (0.007216)
  test_num:
    ["2023-04-05 11:01:35 +0900 [info]: plugin:out_flowcounter_simple\tcount:30\tindicator:num\tunit:second\n"]
    .: (0.111274)
...
```

```diff
diff --git a/test/plugin/test_filter_flowcounter_simple.rb b/test/plugin/test_filter_flowcounter_simple.rb
index aa3ebc1..9635786 100644
--- a/test/plugin/test_filter_flowcounter_simple.rb
+++ b/test/plugin/test_filter_flowcounter_simple.rb
@@ -27,6 +27,7 @@ class FlowCounterSimpleFilterTest < Test::Unit::TestCase
     d = create_driver
     filtered, logs = filter(d, msgs)
     assert_equal msgs, filtered
+    pp logs
     assert { logs.any? { |log| log.include?("count:20") } }
   end

diff --git a/test/plugin/test_out_flowcounter_simple.rb b/test/plugin/test_out_flowcounter_simple.rb
index 2063073..a6fc472 100644
--- a/test/plugin/test_out_flowcounter_simple.rb
+++ b/test/plugin/test_out_flowcounter_simple.rb
@@ -40,6 +40,7 @@ class FlowCounterSimpleOutputTest < Test::Unit::TestCase
       end
       d1.instance.flush_emit(60)
     end
+    pp d1.logs
     assert { d1.logs.any? { |log| log.include?("count:30") } }
   end

@@ -53,6 +54,7 @@ class FlowCounterSimpleOutputTest < Test::Unit::TestCase
       end
       d1.instance.flush_emit(60)
     end
+    pp d1.logs
     assert { d1.logs.any? { |log| log =~ /count:\d+\tindicator:byte\tunit:second/ } }
   end

@@ -66,6 +68,7 @@ class FlowCounterSimpleOutputTest < Test::Unit::TestCase
       end
       d1.instance.flush_emit(60)
     end
+    pp d1.logs
     assert { d1.logs.any? { |log| log.include?("comment:foobar") } }
   end
 end
```

kou replied:
How about removing timestamp and using `assert_equal`?

```ruby
def normalize_logs(logs)
  logs.collect do |log|
    log.gsub(/\A\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2} [-+]\d{4} \[info\]: /, "") 
  end
end

assert_equal(["plugin:filter_flowcounter_simple\tcount:20\tindicator:num\tunit:second\n"],
             normalize_logs(d1.logs))
```

kou replied:
It will show better output (diff of the expected and the actual) on failure.

daipom replied:
Thanks! Indeed!
I have fixed this.

kou commented:
Could you improve the title of this pull request something like `test: don't use deprecated capture_log`?

daipom commented:
> Could you improve the title of this pull request something like `test: don't use deprecated capture_log`?

Thanks! That is more specific and easy to understand. Fixed.

daipom commented:
I just rebased this branch to the latest master.

daipom commented:
I thought that `capture_log` was the method of Fluentd, but it was wrong.

* https://github.com/fluent/fluentd/blob/6c649d189b5cf65ccfce58f6952a31b24c775e72/lib/fluent/test/helpers.rb#L110-L122

This method is defined by this plugin in `helper.rb`.

https://github.com/fluent-plugins-nursery/fluent-plugin-flowcounter-simple/blob/5081e88f97156e019115cab2b4cd810c28ca2ed0/test/helper.rb#L12-L28

But this is the same as the Fluentd method and should not be used.

I changed the title a little, and also fixed the `helper.rb`.

kou commented on test/helper.rb:
```diff
@@ -1,28 +1,13 @@
-require 'rubygems'
-require 'bundler'
+$LOAD_PATH.unshift(File.expand_path("../../", __FILE__))
```
Do we need this?

kou replied:
I think that we don't need this.

It's for requiring files under `test/` such as `test/xxx.rb` by `require "xxx"`.
I think that we should use `require_relative "xxx"` instead of modifying `$LOAD_PATH` global variable.
And we always use `require_relative "../helper"` in `test/**/test_*.rb`.

daipom replied:
Thanks!
It seems that we should fix the template too!

There is still something I don't understand.

> It's for requiring files under `test/` such as `test/xxx.rb` by `require "xxx"`.
> I think that we should use `require_relative "xxx"` instead of modifying `$LOAD_PATH` global variable.
> And we always use `require_relative "../helper"` in `test/**/test_*.rb`.

In my local repository: `/test/fluent-plugin-flowcounter-simple`, the paths in `test/helper.rb` are as follows.

* `__FILE__`: `"/test/fluent-plugin-flowcounter-simple/test/helper.rb"`
* `File.expand_path("../../", __FILE__)`: `"/test/fluent-plugin-flowcounter-simple"`

So this doesn't do anything to enable requiring files under `test/`.
(It enables requiring files directly under the repository's root.)

Since we have the following process, we can use `require helper` without using `$LOAD_PATH`.

https://github.com/fluent-plugins-nursery/fluent-plugin-flowcounter-simple/blob/5081e88f97156e019115cab2b4cd810c28ca2ed0/Rakefile#L5-L6

The same is true for the template.

* https://github.com/fluent/fluentd/blob/v1.16.0/templates/new_gem/Rakefile#L7

This is from the beginning, and I can't find the reason from the following.

* https://github.com/fluent/fluentd/issues/1376
* https://github.com/fluent/fluentd/commit/da3664e44ff5974d8f8097094884edd10a8bc41a#diff-3167ac8636d753c8cc2208b3a3dd46770e00d23ab1b4fa483ddd275c840fa41e

Anyways, it doesn't seem necessary for this plugin, so I will remove this.

kou replied:
OK. Then it's just a bug of the template.

daipom replied:
> I can't find the reason why the template fixes `$LOAD_PATH`.

Perhaps this is just for enabling using relative paths from the repository's root such as `require test/xxx`.

So the conclusion is this after all.

> I think that we should use `require_relative "xxx"` instead of modifying `$LOAD_PATH` global variable.

daipom replied:
> OK. Then it's just a bug of the template.

Yes, we need to consider improving the template.

First, I will remove this from this plugin. Thanks!

daipom replied:
Fixed at https://github.com/fluent-plugins-nursery/fluent-plugin-flowcounter-simple/compare/55b4cf4093e976a26648c869f4060012e6f9a210..515d2d6fec584370fe39e9a4cebeba68293903b6

kou commented on test/helper.rb:
```diff
@@ -1,28 +1,13 @@
-        log.out = tmp
-      end
-    end
+def normalize_logs(logs)
+  logs.collect do |log|
+    log.gsub(/\A\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2} [-+]\d{4} \[info\]: /, "") 
   end
 end
```
We should not define a method in top level.

We should define a helper module, define helper methods in it and include the helper module in each our `*Test`.

daipom replied:
Oh, thanks for the advice.
I will fix it.

daipom replied:
Fixed.

https://github.com/fluent-plugins-nursery/fluent-plugin-flowcounter-simple/compare/e5dc5e920f7d7862be9574fba60d280726fed513..55b4cf4093e976a26648c869f4060012e6f9a210

kou commented on test/helper.rb:
```diff
@@ -1,28 +1,13 @@
-require 'test/unit'
-require 'test/unit/rr'
+Test::Unit::TestCase.include(Fluent::Test::Helpers)
+Test::Unit::TestCase.extend(Fluent::Test::Helpers)
```
Do we need this?

daipom replied:
We don't need this and we can remove this.

I made this change for the same reason as `$LOAD_PATH`.

Is it preferable to remove unnecessary codes rather than keep them aligned with the template?
If so, I will remove these lines.
(I feel that both have advantages and disadvantages, so I cannot judge.)

kou replied:
https://github.com/fluent/fluentd/commit/a5fe470c201dd0fa60b74c5de407bf3bbf96f0e0 said that it's for `config_element`. I think that this is not a good design because unrelated methods such as `assert_equal_event_time`.

I think that we should improve this API in Fluentd itself.

So I think that we don't need this line in this plugin.

daipom replied:
Thanks! I see!
I will remove these lines!

daipom replied:
Fixed at https://github.com/fluent-plugins-nursery/fluent-plugin-flowcounter-simple/compare/55b4cf4093e976a26648c869f4060012e6f9a210..515d2d6fec584370fe39e9a4cebeba68293903b6

daipom commented:
I have removed unnecessary codes from `test/helper.rb`.

kou reviewed (approved):
+1

daipom commented:
Thanks for your review!
